### PR TITLE
I10n pack

### PR DESCRIPTION
**Pull request by sebastienros**
@hishamco BOOM!

### CODE REVIEW

**Comment by sebastienros** [comment]
And now you can see your script running, and you can push changes to fix it

**Comment by hishamco** [comment]
I thought I will work isolated in my branch, anyhow I will keep pushing here until it's done

Thanks a lot

**Comment by sebastienros** [comment]
> I will work isolated in my branch

You are still doing it, it's just that there is a PR that lets you ensure that your script actually works with the CI

**Comment by hishamco** [comment]
packages

@sebastienros Nooooooooooooooooooooooooooooooooooo, this what made me crazy, I was sure, that my script work well, last thing I asked myself to have a look to `Localization` folder

@agriffard Why these files are there?!!

**Comment by hishamco** [comment]
packages

Looks good, but

- `$env:APPVEYOR_BUILD_NUMBER` didn't work nice, it show a number not a semantic version
- Please check Messages tab, there's errors like **NuGet package cannot be read: Package does not contain a manifest.**

Could you please to install one of those packages in a dummy project for verification & validation?

File: Generate-TranslationNuGetPackages.ps1 [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+{
+    $packageName = $packageNamePrefix + $culture;
+    $NuGetSpecFileName = $packageName + $NuGetSpecExtension;
+    $NuGetSpecFilePath = [IO.Path]::Combine($PWD, $localizationFolder, $culture, $NuGetSpecFileName);
+
+    $nugetSpec = [xml](Get-Content -Path $NuGetSpecTemplate);
+    $metadata = $nugetSpec.package.metadata;
+    $metadata.id = $packageName;
+    $metadata.version = $PackageVersionNumber;
+    $metadata.authors = "Hisham Bin Ateya";
```
Comment (hishamco) [inline Generate-TranslationNuGetPackages.ps1]: What the authors & owners should be?

Comment (agriffard) [reply Generate-TranslationNuGetPackages.ps1]: The Orchard Team

File: Generate-TranslationNuGetPackages.ps1 [hunk 2]
```diff
@@ -0,0 +1,76 @@
+﻿$localizationFolder = "Localization";
```
Comment (sebastienros) [inline Generate-TranslationNuGetPackages.ps1]: Is this file necessary since you seem to only use `pack.ps1` ?

File: Generate-TranslationNuGetPackages.ps1 (hunk 1 above)
Comment (sebastienros) [reply Generate-TranslationNuGetPackages.ps1]: Can it be in the template directly? same for the version number and we can just inject the build number in this template?

File: Localization/hu/OrchardCore.Translations.hu.nuspec [hunk 3]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (sebastienros) [inline Localization/hu/OrchardCore.Translations.hu.nuspec]: Should this file be deleted/

File: NuGetSpecTemplate copy.nuspec [hunk 4]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (sebastienros) [inline NuGetSpecTemplate copy.nuspec]: Should this file be deleted?

File: pack.ps1 [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    Remove-Item -Path $artifactsFolder\* -Recurse -Force
+}
+
+New-Item -Path $packagesFolder -ItemType "Directory"
+
+foreach($cultureFolder in $(Get-ChildItem $localizationFolder -Directory)) {
+    createNuGetPackage $cultureFolder.Name
+}
+
+Move-Item -Path $packagesFolder -Destination $artifactsFolder
```
Comment (sebastienros) [inline pack.ps1]: Looks like you could only use the artifactsFolder directly, not have the package intermediate packagesFolder.

Comment (hishamco) [reply pack.ps1]: You are right, I did that while I'm testing why pot files become nupkg files

File: Generate-TranslationNuGetPackages.ps1 (hunk 1 above)
Comment (hishamco) [reply Generate-TranslationNuGetPackages.ps1]: Of course it will be in the template, regarding the version I'm still thinking about better way if there's

**Comment by sebastienros** [comment]
You don't need the nuget cli to push packages , appveyor will do it by itself, once I configure the API key ( which is not right now on purpose)

**Comment by hishamco** [comment]
> You don't need the nuget cli to push packages , appveyor will do it by itself

Yep, I know that, I just used nuget cli to create the `.nupkg` instead of doing it myself, because I faces some issues with psmdcp hash, so now all errors is off and everything is work fine

I'm working offline to to make the script pretty, hope to push the updates soon ..

**Comment by sebastienros** [comment]
Makes sense, thanks

**Comment by hishamco** [comment]
@sebastienros I think it's time to :shipit:

File: appveyor.yml [hunk 6]
```diff
@@ -5,12 +5,19 @@ branches:
...
+environment:
+  packageVersion: 1.0.0
 init:
   - git config --global core.autocrlf true
+  - ps: $env:buildVersion = "$env:packageVersion.$env:appveyor_build_number"
+  - ps: $env:nugetVersion = "$env:packageVersion-beta-$env:appveyor_build_number"
+  - ps: Update-AppveyorBuild -Version $env:buildVersion
 install:
-  - ps: $env:BuildNumber= $env:APPVEYOR_BUILD_NUMBER
+  - ps: Start-FileDownload 'https://dist.nuget.org/win-x86-commandline/latest/nuget.exe'
```
Comment (sebastienros) [inline appveyor.yml]: Where is it used?

File: appveyor.yml [hunk 7]
```diff
@@ -5,12 +5,19 @@ branches:
   only:
     - l10n_master # The branch that Crowdin targets when updates are made on .po files
     - master      # The branch that is published on Nuget.org
+    - l10n-pack   # For testing purpose
+environment:
+  packageVersion: 1.0.0
 init:
   - git config --global core.autocrlf true
+  - ps: $env:buildVersion = "$env:packageVersion.$env:appveyor_build_number"
+  - ps: $env:nugetVersion = "$env:packageVersion-beta-$env:appveyor_build_number"
```
Comment (sebastienros) [inline appveyor.yml]: Why do we need two values for a build number? one with `beta` and one without.

Comment (sebastienros) [inline appveyor.yml]: Can we move the version prefix to its custom variable? (beta)

Comment (hishamco) [reply appveyor.yml]: The `$env:buildVersion` will be used by AppVeyor itself, and `$env:nugetVersion` the one will be used in `.nupkg`

To see `$env:buildVersion` in action, please keep your eyes on the top right while AppVeyor is running, the version will be something like `1.0.0.54` instead of `1.0.54`

Comment (hishamco) [reply appveyor.yml]: We can move the `-beta-$env:appveyor_build_number` to something like $prelease or $qualifier, but I don't think this will add a value to the `pack.ps`

Comment (sebastienros) [reply appveyor.yml]: You don't answer the `why` in my question. I don't think this is necessary. What would happen if you didn't call `Update-AppveyorBuild`? IMO it would still have a version, as this is defined in the template :

image

The only difference is that you add an extra `.0`, but we don't really care. And you really really really want that, I can change the template in the settings.

Comment (sebastienros) [reply appveyor.yml]: It has as much value as the variable `packageVersion: 1.0.0` that you created. And I don't mind either merging these two in a single value.

Comment (hishamco) [reply appveyor.yml]: If you change the template it will be YES, nothing but the new environment variable that I introduced `packageVersion` will take care of this, so whenever we change the version it will change both the AppVeyor & NuGet version, two birds with one stone

Hope it's answer the question

Comment (sebastienros) [reply appveyor.yml]: Or, if you really want to give a meaningful version to appveyor (for some reason I don't get) then we could use the correct one, which is in `env:nugetVersion`

Comment (hishamco) [reply appveyor.yml]: In this case both are equivalent, but I differentiate between them because sometimes we may need to add preleases or qualifier to the package name, but this is not the case for the AppVeyor

Comment (hishamco) [reply appveyor.yml]: As I said before if we will no longer use preleases or qualifier in versioning it's fine to use only one variable

Comment (sebastienros) [reply appveyor.yml]: > but this is not the case for the AppVeyor

Right, because we really don't care about the appveyor version, which is why I don't get why you even want to update it.

Comment (hishamco) [reply appveyor.yml]: If we don't care I will modify it right now

Comment (hishamco) [reply appveyor.yml]: BOOM it's ready

File: appveyor.yml [hunk 8]
```diff
@@ -8,9 +8,10 @@ branches:
     - l10n-pack   # For testing purpose
 environment:
   packageVersion: 1.0.0
+  versionPrefix: -beta
```
Comment (hishamco) [inline appveyor.yml]: We could add it to the `packageVersion`

File: appveyor.yml (hunk 6 above)
Comment (sebastienros) [reply appveyor.yml]: Closed the conversation by mistake

Comment (sebastienros) [reply appveyor.yml]: Sorry, found it

File: appveyor.yml [hunk 9]
```diff
@@ -5,12 +5,18 @@ branches:
   only:
     - l10n_master # The branch that Crowdin targets when updates are made on .po files
     - master      # The branch that is published on Nuget.org
+    - l10n-pack   # For testing purpose
```
Comment (hishamco) [inline appveyor.yml]: This branch was for testing purpose only, I think it's fine to remove it, but I'm not sure if the CI/CD will kicks in
